app/src/lib/is-git-on-path.ts

```diff
@@ -1,47 +1,75 @@
+    const cp = spawn(command, args, options)
 
-  // adapted from http://stackoverflow.com/a/34953561/1363815
-  if (__WIN32__) {
-    return new Promise<boolean>((resolve, reject) => {
-      const windowsRoot = process.env.SystemRoot || 'C:\\Windows'
-      const wherePath = Path.join(windowsRoot, 'System32', 'where.exe')
+    cp.on('error', error => {
+      log.warn(`Unable to spawn ${command}`, error)
+      resolve(undefined)
+    })
 
-      const cp = spawn(wherePath, ['git'])
+    const chunks = new Array<Buffer>()
+    let total = 0
 
-      cp.on('error', error => {
-        log.warn('Unable to spawn where.exe', error)
-        resolve(false)
-      })
+    cp.stdout.on('data', (chunk: Buffer) => {
+      chunks.push(chunk)
+      total += chunk.length
+    })
 
-      // `where` will return 0 when the executable
-      // is found under PATH, or 1 if it cannot be found
-      cp.on('close', function(code) {
-        resolve(code === 0)
-      })
-      return
+    cp.on('close', function(code) {
+      if (code !== 0) {
+        resolve(undefined)
+      } else {
+        resolve(
+          Buffer.concat(chunks, total)
+            .toString()
+            .replace(/\r?\n[^]*/m, '')
+        )
+      }
     })
+  })
+}
+
+export function findGitOnPath(): Promise<string | undefined> {
+  // adapted from http://stackoverflow.com/a/34953561/1363815
+  if (__WIN32__) {
+    const windowsRoot = process.env.SystemRoot || 'C:\\Windows'
+    const wherePath = Path.join(windowsRoot, 'System32', 'where.exe')
+
+    // `where` will list _all_ PATH components where the executable
+    // is found, one per line, and return 0, or print an error and
+    // return 1 if it cannot be found
+    log.info(`calling captureCommandOutput(where git)`)
+    return captureCommandOutput(wherePath, ['git'], { cwd: windowsRoot })
+  }
+
+  if (__DARWIN__ || __LINUX__) {
+    // `which` will print the path and return 0 when the executable
+    // is found under PATH, or return 1 if it cannot be found
+    return captureCommandOutput('which', ['git'])
   }
 
-  if (__LINUX__) {
-    return new Promise<boolean>((resolve, reject) => {
-      const process = spawn('which', ['git'])
+  return Promise.resolve(undefined)
+}
 
-      // `which` will return 0 when the executable
-      // is found under PATH, or 1 if it cannot be found
-      process.on('close', function(code) {
-        resolve(code === 0)
-      })
-    })
+export function isGitOnPath(): Promise<boolean> {
+  // Modern versions of macOS ship with a Git shim that guides you through
+  // the process of setting everything up. We trust this is available, so
+  // don't worry about looking for it here.
+  if (__DARWIN__) {
+    return Promise.resolve(true)
   }
 
-  return Promise.resolve(false)
+  return findGitOnPath().then(
+    (path: string | undefined) =>
+      new Promise<boolean>((resolve, reject) => {
+        resolve(path !== undefined)
+      })
+  )
```

Comment:
we use `async`/`await` pretty liberally in the desktop codebase, so i think this could be simplified to
```suggestion
  return (await findGitOnPath()) !== undefined
```

Reply:
Right, I would prefer that, too, but as I was only refactoring, I tried to keep the style as close to the original as possible: https://github.com/desktop/desktop/blob/1123057da3728015853bb8ac859e296575bc4ef7/app/src/lib/is-git-on-path.ts#L14-L31

If you are okay with changing the style while refactoring, I will be delighted to make that change!

Reply:
Oh right, since you suggested that change, I figure this project does not want the commits to be rebased and squashed, so I'll just accept your suggestion. Thanks!